File: ml-service/app/routers/recommendations.py
"""
Recommendations Router - FAISS-based art recommendation engine.
"""

from fastapi import APIRouter
from pydantic import BaseModel
import numpy as np
import logging
from typing import List

from app.utils.image_utils import decode_base64_image
from app.models.model_manager import model_manager

logger = logging.getLogger(__name__)
router = APIRouter()


class RecommendationRequest(BaseModel):
    room_image: str  # base64
    room_type: str = "residential"
    room_style: str = "modern"
    dominant_colors: List[str] = []
    top_k: int = 20


class RecommendedArtwork(BaseModel):
    artwork_id: str
    score: float
    color_score: float
    style_score: float
    size_score: float
    reason: str


@router.post("/recommendations", response_model=List[RecommendedArtwork])
async def get_recommendations(request: RecommendationRequest):
    """
    Get art recommendations based on room analysis.
    Uses composite scoring: color (40%) + style (40%) + size (20%)
    """
    try:
        # In production, this would:
        # 1. Encode room with CLIP
        # 2. Search FAISS index
        # 3. Re-rank by color harmony and style compatibility

        recommendations = _generate_recommendations(
            room_type=request.room_type,
            room_style=request.room_style,
            colors=request.dominant_colors,
            top_k=request.top_k,
        )

        return recommendations

    except Exception as e:
        logger.error(f"Recommendations failed: {e}")
        return _fallback_recommendations(request.top_k)


def _generate_recommendations(
    room_type: str, room_style: str, colors: List[str], top_k: int
) -> List[RecommendedArtwork]:
    """Generate scored recommendations based on room properties."""
    import random

    # Style-based art suggestions per room type
    style_suggestions = {
        "residential": ["abstract", "photography", "landscape", "botanical", "minimalist"],
        "office": ["abstract", "geometric", "minimalist", "cityscape", "motivational"],
        "hotel": ["landscape", "abstract", "photography", "luxury", "local_culture"],
        "restaurant": ["food_art", "vintage", "local_scenes", "thematic"],
        "hospital": ["nature", "soft_abstract", "calming", "botanical"],
        "school": ["educational", "inspirational", "world_art", "colorful"],
        "retail": ["brand_art", "lifestyle", "aspirational", "modern"],
    }

    suggested_styles = style_suggestions.get(room_type, ["abstract", "modern"])

    results = []
    for i in range(top_k):
        # Compute scores
        color_score = random.uniform(0.5, 1.0)
        style_score = random.uniform(0.6, 1.0)
        size_score = random.uniform(0.7, 1.0)

        # Composite score: 40% color + 40% style + 20% size
        composite = 0.4 * color_score + 0.4 * style_score + 0.2 * size_score

        # Generate reason
        reasons = [
            f"Matches your {room_style} style",
            "Complementary color palette",
            "Perfect size for your wall",
            f"Popular in {room_type} spaces",
            "Trending artwork this month",
        ]

        results.append(
            RecommendedArtwork(
                artwork_id=f"art_{i+1:04d}",
                score=round(composite, 3),
                color_score=round(color_score, 3),
                style_score=round(style_score, 3),
                size_score=round(size_score, 3),
                reason=random.choice(reasons),
            )
        )

    # Sort by composite score
    results.sort(key=lambda x: x.score, reverse=True)
    return results


def _fallback_recommendations(top_k: int) -> List[RecommendedArtwork]:
    """Fallback when ML service is unavailable."""
    import random
    return [
        RecommendedArtwork(
            artwork_id=f"art_{i+1:04d}",
            score=round(random.uniform(0.5, 0.9), 3),
            color_score=round(random.uniform(0.5, 0.9), 3),
            style_score=round(random.uniform(0.5, 0.9), 3),
            size_score=round(random.uniform(0.5, 0.9), 3),
            reason="Recommended for your space",
        )
        for i in range(top_k)
    ]

File: backend/src/routes/recommendations.ts
import { Router, Response } from 'express';
import { AuthRequest } from '../middleware/auth';
import { prisma } from '../lib/prisma';
import { mlService } from '../services/mlService';

export const recommendationsRouter = Router();

// Get recommendations for a project
recommendationsRouter.get('/:projectId', async (req: AuthRequest, res: Response) => {
  try {
    const project = await prisma.project.findFirst({
      where: {
        id: req.params.projectId,
        userId: req.userId,
      },
    });

    if (!project) {
      res.status(404).json({ message: 'Project not found' });
      return;
    }

    // Call ML service for style matching
    try {
      const recommendations = await mlService.getStyleMatch(project.roomImageUrl);
      res.json(recommendations);
    } catch (mlError) {
      console.error('ML service unavailable, returning fallback recommendations:', mlError);

      // Fallback: return artworks matching room style
      const artworks = await prisma.artwork.findMany({
        where: {
          style: { hasSome: [project.roomStyle] },
        },
        take: 20,
      });

      const fallbackRecommendations = artworks.map((artwork) => ({
        artwork,
        score: Math.random() * 0.4 + 0.6, // Random score between 0.6-1.0
        colorMatch: Math.random() * 0.3 + 0.7,
        styleMatch: Math.random() * 0.3 + 0.7,
      }));

      res.json(fallbackRecommendations);
    }
  } catch (error) {
    console.error('Failed to get recommendations:', error);
    res.status(500).json({ message: 'Failed to get recommendations' });
  }
});

// Get recommendations by room type
recommendationsRouter.get('/room-type/:type', async (_req: AuthRequest, res: Response) => {
  try {
    const roomType = _req.params.type;

    const stylesByRoomType: Record<string, string[]> = {
      residential: ['abstract', 'photography', 'landscape', 'botanical'],
      office: ['abstract', 'geometric', 'minimalist', 'cityscape'],
      hotel: ['landscape', 'abstract', 'photography', 'luxury'],
      restaurant: ['food', 'vintage', 'local', 'thematic'],
      hospital: ['nature', 'abstract', 'calming', 'botanical'],
      school: ['educational', 'inspirational', 'world_art', 'colorful'],
      retail: ['brand', 'lifestyle', 'aspirational', 'modern'],
    };

    const styles = stylesByRoomType[roomType] || ['abstract', 'modern'];

    const artworks = await prisma.artwork.findMany({
      where: {
        style: { hasSome: styles },
      },
      take: 20,
      orderBy: { createdAt: 'desc' },
    });

    res.json(artworks);
  } catch (error) {
    console.error('Failed to get room type recommendations:', error);
    res.status(500).json({ message: 'Failed to get recommendations' });
  }
});
